presto-main/src/main/java/com/facebook/presto/execution/PlanFlattener.java

```diff
@@ -0,0 +1,245 @@
+        {
+            ImmutableList.Builder<JsonNode> nodes = ImmutableList.builder();
+            fragment.getRoot().accept(visitor, nodes);
+            return new FlattenedPlanFragment(explainer.getPlan(fragment, session), fragment, nodes.build());
+        }
+
+        @JsonProperty
+        public PlanFragmentId getId()
+        {
+            return fragment.getId();
+        }
+
+        @JsonProperty
+        public String getTextPlan()
+        {
+            return textPlan;
+        }
+
+        @JsonProperty
+        public PlanNode getRoot()
+        {
+            return fragment.getRoot();
+        }
+
+        @JsonProperty
+        public List<JsonNode> getNodes()
+        {
+            return nodes;
+        }
+
+        @JsonProperty
+        public Map<Symbol, Type> getSymbols()
+        {
+            return fragment.getSymbols();
+        }
+
+        @JsonProperty
+        public PartitioningHandle getPartitioning()
+        {
+            return fragment.getPartitioning();
+        }
+
+        @JsonProperty
+        public List<PlanNodeId> getPartitionedSources()
+        {
+            return fragment.getPartitionedSources();
+        }
+
+        @JsonProperty
+        public List<Type> getTypes()
+        {
+            return fragment.getTypes();
+        }
+
+        @JsonProperty
+        public Set<PlanNode> getPartitionedSourceNodes()
+        {
+            return fragment.getPartitionedSourceNodes();
+        }
+
+        @JsonProperty
+        public List<RemoteSourceNode> getRemoteSourceNodes()
+        {
+            return fragment.getRemoteSourceNodes();
+        }
+
+        @JsonProperty
+        public PartitioningScheme getPartitioningScheme()
+        {
+            return fragment.getPartitioningScheme();
+        }
+    }
+
+    private static class FlatteningVisitor
+            extends SimplePlanVisitor<ImmutableList.Builder<JsonNode>>
+    {
+        private final ObjectMapper flatteningMapper;
+
+        public FlatteningVisitor(ObjectMapper originalMapper)
+        {
+            this.flatteningMapper = originalMapper.copy();
+
+            PlanNodeFlatteningSerializer flattener = new PlanNodeFlatteningSerializer(originalMapper.getSerializerProviderInstance());
+            flatteningMapper.registerModule(new SimpleModule().addSerializer(PlanNode.class, flattener));
+        }
+
+        @Override
+        protected Void visitPlan(PlanNode node, ImmutableList.Builder<JsonNode> context)
+        {
+            context.add(flatteningMapper.valueToTree(node));
+            return super.visitPlan(node, context);
+        }
+    }
+
+    /**
+     * The flattening serializer uses different techniques to serialize the root (i.e. the node it's passed) and
+     * that node's children. For the root, it uses the subtype serializer (e.g. OutputNode), for children it uses
+     * the supertype serializer (i.e. the serializer for PlanNode). This ensures that the only properties that
```

Comment:
put the last sentence first so we understand the point before getting into the details